firewall/inspection/portscan/detect.go

```diff
@@ -0,0 +1,357 @@
+package portscan
+
+import (
+	"context"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/safing/portbase/log"
+	"github.com/safing/portbase/modules"
+	"github.com/safing/portmaster/firewall/inspection"
+	"github.com/safing/portmaster/netenv"
+	"github.com/safing/portmaster/network"
+	"github.com/safing/portmaster/network/netutils"
+	"github.com/safing/portmaster/network/packet"
+	"github.com/safing/portmaster/process"
+	"github.com/safing/portmaster/status"
+)
+
+type tcpUDPport struct {
+	protocol packet.IPProtocol
+	port     uint16
+}
+
+type ipData struct {
+	score int //score needs to be big enough to keep maxScore + addScore... to prevent overflow
+	//	greylistingWorked bool
+	previousOffender bool
+	blocked          bool
+	ignore           bool
+	lastSeen         time.Time
+	lastUpdated      time.Time
+	blockedPorts     []tcpUDPport
+}
+
+const (
+	//fixme: Which production-values do we want to have?
+	cleanUpInterval = 1 * time.Minute //fixme: Debug-Value
+	cleanUpMaxDelay = 5 * time.Minute
+
+	startAfter            = 1 * time.Second //fixme: Debug Value; When should the Portscan Detection start to prevent blocking Apps that just try to reconnect?
+	decreaseInterval      = 11 * time.Second
+	unblockIdleTime       = 1 * time.Hour
+	undoSuspicionIdleTime = 24 * time.Hour
+	unignoreTime          = 24 * time.Hour
+
+	startRegisteredPorts = 1024
+	startDynamicPorts    = 32768
+
+	addScoreWellKnownPort  = 40
+	addScoreRegisteredPort = 20
+	addScoreDynamicPort    = 10
+
+	scoreBlock = 160
+	maxScore   = 320
+
+	threadPrefix = "portscan: "
+)
+
+var (
+	ips    map[string]*ipData
+	ownIPs []net.IP
+
+	module     *modules.Module
+	runOnlyOne sync.Mutex
+)
+
+// Detector detects if a connection is part of a portscan which already sent some packets.
+type Detector struct{}
+
+// Name implements the inspection interface.
+func (d *Detector) Name() string {
+	return "Portscan Detection"
+}
+
+// Inspect implements the inspection interface.
+func (d *Detector) Inspect(conn *network.Connection, pkt packet.Packet) (pktVerdict network.Verdict, proceed bool, err error) {
+	runOnlyOne.Lock()
+	defer runOnlyOne.Unlock()
+
+	ctx := pkt.Ctx()
+
+	//Delete for production. This just reduces the amount of Debug Messages significantly
+	//	if conn.LocalIP.Equal(net.IP([]byte{255, 255, 255, 255})) {
+	//		return network.VerdictUndecided, false, nil
+	//	}
+	log.Tracer(ctx).Debugf("new connection for Portscan detection")
+
+	rIP, ok := conn.Entity.GetIP() //remote IP
+	if !ok {                       //No IP => return undecided
+		return network.VerdictUndecided, false, nil
+	}
+
+	ipString := rIP.String()
+	entry, inMap := ips[ipString]
+
+	log.Tracer(ctx).Debugf("Conn: %v, Entity: %#v, Protocol: %v, LocalIP: %s, LocalPort: %d, inMap: %v, entry: %+v", conn, conn.Entity, conn.IPProtocol, conn.LocalIP.String(), conn.LocalPort, inMap, entry)
```

Comment:
Thanks!

I actually expected log.Tracer().Debugf() to not execute the Sprintf if the loglevel isn't set to Debug or Trace. Then it would only have impacted Debggging-Mode. Not sure, but it seems as I was wrong.

Removed most instances of %v (exceptions: 1x Error-handling, 1x: Printing slice of own IPs)

Reply:
:+1:

You may be right that the current log does not evalute it (hint: [it does](https://github.com/safing/portbase/blob/develop/log/trace.go#L178), but [will soon stop](https://github.com/safing/portbase/blob/feature/gomod/log/input.go#L11) doing it). 

In my experience logging is something that can quickly be changed (because one uses a different library or adds features like log-aggregation, ...) so relying on it to not evaluate format strings is dangerous. You may only display warning on the screen but write all traces to a file (i.e. different log levels). So better avoid using those format-strings specifiers in hot paths :+1: